packages/codemirror-blocks/src/state/actions.ts

```diff
@@ -83,11 +83,11 @@ export const setCode = (code: string) => ({ type: "SET_CODE" as const, code });
 export const setBlockMode =
   (
     enabled: boolean,
-    editor: ReadonlyRangedText,
+    code: string,
     language: Language
   ): AppThunk<ReturnType<typeof checkASTRoundtripConversion>> =>
   (dispatch) => {
-    const result = checkASTRoundtripConversion(editor.getValue(), language);
```

Comment:
I'm curious - why does `setBlockMode()` need to receive the editor contents? Isn't the code already a prop of ToggleEditor?

Reply:
Well, for one, this is a redux action, and redux actions only have access to the redux store and the parameters passed to the action. They don't know or care about react.

The redux store holds state for "code", but this is only when in block mode, and is just the pretty printed version of the AST. When not in block mode, the code that's in codemirror doesn't get synced back to the store. So if you are in text mode and switch to block mode, the only way for the action to know what code to convert is for it to be passed in. We could sync text content of the editor with the redux store on every key press, but we'll have to be careful with performance to make sure we don't rerender the whole world every time.